File: go/fingerprint.go
// Package preflight is the canonical, independently publishable Go
// implementation of IFF's x402 requirement fingerprint and preflight API.
// The private monitor imports this module for production fingerprinting, so
// the publicly reviewed algorithm and the deployed algorithm do not drift.
//
// The full protocol this client wraps -- the fingerprint algorithm, the
// Merkle transparency log, signed tree heads, and inclusion/consistency
// proofs -- is specified in ../spec/x402-requirement-transparency-v1.md,
// written so a third party can implement a compatible verifier without
// reading this SDK's source. ../spec/verify_example.py is a dependency-light,
// runnable reference verifier that checks a
// live log directly against that spec, independent of this SDK.
//
// This module's canonical, public home is
// https://github.com/ifandonlyif-io/iff-x402-transparency. During extraction
// the private monitor uses a local replace directive; after the first public
// tag it will depend on that immutable tagged version instead.
package preflight

import (
	"bytes"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"regexp"
	"sort"
	"strings"
)

// FingerprintVersion is the current C1 requirement-fingerprint version.
const FingerprintVersion = 1

const (
	optionFingerprintDomain = "iff-x402-option/v1\n"
	setFingerprintDomain    = "iff-x402-set/v1\n"
	payeeFingerprintDomain  = "iff-x402-payee/v1\n"
)

// evmAddressPattern matches a lowercase-or-mixed-case 20-byte hex address.
// Only values matching this pattern are safe to lowercase; base58 addresses
// (Solana and others) are case-sensitive and must pass through unchanged.
var evmAddressPattern = regexp.MustCompile(`^0x[0-9a-fA-F]{40}$`)

// PaymentOption is the x402 v2 payment option shape this SDK fingerprints:
// the same safe subset the public v3 evidence API carries.
type PaymentOption struct {
	Scheme            string `json:"scheme"`
	Network           string `json:"network"`
	Asset             string `json:"asset"`
	Amount            string `json:"amount"`
	PayTo             string `json:"pay_to"`
	MaxTimeoutSeconds int    `json:"max_timeout_seconds"`
}

// Fingerprint identifies a normalized payment-option set.
type Fingerprint struct {
	Version   int
	SetFP     string
	OptionFPs []string
}

// PayeeFingerprint is C4 rule 4b's amount-blind payment identity.
type PayeeFingerprint struct {
	Version    int
	PayeeSetFP string
	PayeeFPs   []string
}

type canonicalPaymentOption struct {
	Scheme  string `json:"scheme"`
	Network string `json:"network"`
	Asset   string `json:"asset"`
	PayTo   string `json:"pay_to"`
	Amount  string `json:"amount"`
}

type canonicalPayeeOption struct {
	Scheme  string `json:"scheme"`
	Network string `json:"network"`
	Asset   string `json:"asset"`
	PayTo   string `json:"pay_to"`
}

// ComputeFingerprint derives the C1 requirement fingerprint for a set of
// x402 payment options. ok is false when options is empty.
func ComputeFingerprint(options []PaymentOption) (Fingerprint, bool) {
	if len(options) == 0 {
		return Fingerprint{}, false
	}
	seen := make(map[string]struct{}, len(options))
	optionFPs := make([]string, 0, len(options))
	for _, option := range options {
		fp := computeOptionFingerprint(option)
		if _, duplicate := seen[fp]; duplicate {
			continue
		}
		seen[fp] = struct{}{}
		optionFPs = append(optionFPs, fp)
	}
	sort.Strings(optionFPs)
	return Fingerprint{
		Version:   FingerprintVersion,
		SetFP:     computeSetFingerprint(optionFPs),
		OptionFPs: optionFPs,
	}, true
}

// ComputePayeeFingerprint derives C4 rule 4b's amount-blind payee
// fingerprint: the same options as ComputeFingerprint, hashed without the
// amount field, so two option sets that differ only in price share a
// PayeeSetFP. ok is false when options is empty.
func ComputePayeeFingerprint(options []PaymentOption) (PayeeFingerprint, bool) {
	if len(options) == 0 {
		return PayeeFingerprint{}, false
	}
	seen := make(map[string]struct{}, len(options))
	payeeFPs := make([]string, 0, len(options))
	for _, option := range options {
		fp := computePayeeOptionFingerprint(option)
		if _, duplicate := seen[fp]; duplicate {
			continue
		}
		seen[fp] = struct{}{}
		payeeFPs = append(payeeFPs, fp)
	}
	sort.Strings(payeeFPs)
	return PayeeFingerprint{
		Version:    FingerprintVersion,
		PayeeSetFP: computeSetFingerprint(payeeFPs),
		PayeeFPs:   payeeFPs,
	}, true
}

func computeOptionFingerprint(option PaymentOption) string {
	canonical := canonicalPaymentOption{
		Scheme:  strings.ToLower(strings.TrimSpace(option.Scheme)),
		Network: strings.ToLower(strings.TrimSpace(option.Network)),
		Asset:   NormalizeAddressLikeField(option.Asset),
		PayTo:   NormalizeAddressLikeField(option.PayTo),
		Amount:  NormalizeAmount(option.Amount),
	}
	return domainSeparatedHash(optionFingerprintDomain, marshalCanonicalJSON(canonical))
}

func computePayeeOptionFingerprint(option PaymentOption) string {
	canonical := canonicalPayeeOption{
		Scheme:  strings.ToLower(strings.TrimSpace(option.Scheme)),
		Network: strings.ToLower(strings.TrimSpace(option.Network)),
		Asset:   NormalizeAddressLikeField(option.Asset),
		PayTo:   NormalizeAddressLikeField(option.PayTo),
	}
	return domainSeparatedHash(payeeFingerprintDomain, marshalCanonicalJSON(canonical))
}

func computeSetFingerprint(sortedUniqueFPs []string) string {
	hasher := sha256.New()
	hasher.Write([]byte(setFingerprintDomain))
	hasher.Write([]byte(strings.Join(sortedUniqueFPs, "\n")))
	return hex.EncodeToString(hasher.Sum(nil))
}

func domainSeparatedHash(domain string, payload []byte) string {
	hasher := sha256.New()
	hasher.Write([]byte(domain))
	hasher.Write(payload)
	return hex.EncodeToString(hasher.Sum(nil))
}

// NormalizeAddressLikeField implements the C1 asset/pay_to rule: an
// EVM-style 0x address is lowercased; anything else (in particular a
// case-sensitive base58 address) is only trimmed.
func NormalizeAddressLikeField(value string) string {
	trimmed := strings.TrimSpace(value)
	if evmAddressPattern.MatchString(trimmed) {
		return strings.ToLower(trimmed)
	}
	return trimmed
}

// NormalizeAmount implements the C1 amount rule: trim, then strip leading
// zeros only when the trimmed value is purely ASCII decimal digits.
func NormalizeAmount(value string) string {
	trimmed := strings.TrimSpace(value)
	if trimmed == "" {
		return trimmed
	}
	for _, r := range trimmed {
		if r < '0' || r > '9' {
			return trimmed
		}
	}
	stripped := strings.TrimLeft(trimmed, "0")
	if stripped == "" {
		return "0"
	}
	return stripped
}

// marshalCanonicalJSON produces compact, HTML-unescaped JSON, matching C1's
// contract (encoding/json.Marshal always HTML-escapes '<'/'>'/'&', which C1
// forbids).
func marshalCanonicalJSON(value interface{}) []byte {
	var buf bytes.Buffer
	encoder := json.NewEncoder(&buf)
	encoder.SetEscapeHTML(false)
	_ = encoder.Encode(value)
	return bytes.TrimRight(buf.Bytes(), "\n")
}

File: ts/src/fingerprint.ts
// Every normalization rule and hash domain here is tested against the same
// conformance vectors as IFF's production evidence service.
//
// SHA-256 is computed via WebCrypto (globalThis.crypto.subtle), available
// without any package dependency in supported Node >=22 releases and in every
// browser. That is the only reason these functions are async: Go's
// ComputeFingerprint is synchronous, but crypto.subtle.digest always returns
// a Promise.

/** The x402 v2 payment option shape this SDK fingerprints. Field names are
 * idiomatic camelCase (matching the x402 wire format's `payTo`/
 * `maxTimeoutSeconds`). The canonical JSON mapping is an implementation
 * detail of canonicalOptionJson/canonicalPayeeJson below. */
export interface PaymentOption {
  scheme: string;
  network: string;
  asset: string;
  amount: string;
  payTo: string;
  maxTimeoutSeconds?: number;
}

/** Canonical IFF fingerprint format version. */
export const FINGERPRINT_VERSION = 1;

const OPTION_FINGERPRINT_DOMAIN = "iff-x402-option/v1\n";
const SET_FINGERPRINT_DOMAIN = "iff-x402-set/v1\n";
// C4 rule 4b's payee-fingerprint option-level domain. The set-level hash
// deliberately reuses SET_FINGERPRINT_DOMAIN unchanged -- see
// computePayeeFingerprint's doc comment for why that is still safe.
const PAYEE_FINGERPRINT_DOMAIN = "iff-x402-payee/v1\n";

// Only a literal lowercase "0x" prefix followed by 40 hex digits is
// case-folded (real EIP-55 checksum addresses always keep a lowercase "0x"
// prefix); anything else, including a base58 address, passes through
// trimmed but otherwise unchanged.
const EVM_ADDRESS_PATTERN = /^0x[0-9a-fA-F]{40}$/;

export interface Fingerprint {
  version: number;
  setFingerprint: string;
  optionFingerprints: string[];
}

export interface PayeeFingerprint {
  version: number;
  payeeSetFingerprint: string;
  payeeFingerprints: string[];
}

/** Implements C1's asset/pay_to normalization rule. */
function normalizeAddressLikeField(value: string): string {
  const trimmed = value.trim();
  return EVM_ADDRESS_PATTERN.test(trimmed) ? trimmed.toLowerCase() : trimmed;
}

/** Implements C1's amount normalization rule: trim, then strip leading
 * zeros only when the trimmed value is purely ASCII decimal digits. */
function normalizeAmount(value: string): string {
  const trimmed = value.trim();
  if (trimmed === "") {
    return trimmed;
  }
  if (!/^[0-9]+$/.test(trimmed)) {
    return trimmed;
  }
  const stripped = trimmed.replace(/^0+/, "");
  return stripped === "" ? "0" : stripped;
}

/** The canonical, fixed-key-order option shape. JS object literals preserve
 * string-key insertion order, and JSON.stringify does not HTML-escape
 * '<'/'>'/'&', so both contract properties require no extra configuration. */
function canonicalOptionJson(option: PaymentOption): string {
  return JSON.stringify({
    scheme: option.scheme.trim().toLowerCase(),
    network: option.network.trim().toLowerCase(),
    asset: normalizeAddressLikeField(option.asset),
    pay_to: normalizeAddressLikeField(option.payTo),
    amount: normalizeAmount(option.amount),
  });
}

/** canonicalOptionJson without amount -- C4 rule 4b's payee identity. */
function canonicalPayeeJson(option: PaymentOption): string {
  return JSON.stringify({
    scheme: option.scheme.trim().toLowerCase(),
    network: option.network.trim().toLowerCase(),
    asset: normalizeAddressLikeField(option.asset),
    pay_to: normalizeAddressLikeField(option.payTo),
  });
}

async function sha256Hex(message: string): Promise<string> {
  const bytes = new TextEncoder().encode(message);
  const digest = await globalThis.crypto.subtle.digest("SHA-256", bytes);
  return Array.from(new Uint8Array(digest))
    .map((byte) => byte.toString(16).padStart(2, "0"))
    .join("");
}

/** Hashes the domain and payload as one byte sequence. */
async function domainSeparatedHash(domain: string, payload: string): Promise<string> {
  return sha256Hex(domain + payload);
}

/** Sorts and deduplicates already-computed per-option fingerprints. */
function dedupeAndSort(fingerprints: string[]): string[] {
  return Array.from(new Set(fingerprints)).sort();
}

/**
 * Derives the canonical requirement fingerprint for a set of x402 payment
 * options. The maxTimeoutSeconds transport field is deliberately excluded
 * from the hashed shape.
 *
 * Returns null when the set has no fingerprintable option (an empty
 * array).
 */
export async function computeFingerprint(options: PaymentOption[]): Promise<Fingerprint | null> {
  if (options.length === 0) {
    return null;
  }
  const optionFingerprints: string[] = [];
  for (const option of options) {
    optionFingerprints.push(await domainSeparatedHash(OPTION_FINGERPRINT_DOMAIN, canonicalOptionJson(option)));
  }
  const sorted = dedupeAndSort(optionFingerprints);
  const setFingerprint = await domainSeparatedHash(SET_FINGERPRINT_DOMAIN, sorted.join("\n"));
  return { version: FINGERPRINT_VERSION, setFingerprint, optionFingerprints: sorted };
}

/**
 * Derives an amount-blind payee fingerprint: the same options as
 * computeFingerprint, hashed without the amount field, so two option sets
 * that differ only in price share the same payeeSetFingerprint.
 *
 * The set-level hash reuses SET_FINGERPRINT_DOMAIN unchanged: it only needs
 * to separate a set-level hash from an option-level hash, which the
 * differing option-level domains (OPTION_FINGERPRINT_DOMAIN vs
 * PAYEE_FINGERPRINT_DOMAIN) already guarantee -- so a payeeSetFingerprint
 * can never collide with an ordinary setFingerprint for the same options.
 *
 * Returns null when the set has no fingerprintable option, mirroring
 * computeFingerprint.
 */
export async function computePayeeFingerprint(options: PaymentOption[]): Promise<PayeeFingerprint | null> {
  if (options.length === 0) {
    return null;
  }
  const payeeFingerprints: string[] = [];
  for (const option of options) {
    payeeFingerprints.push(await domainSeparatedHash(PAYEE_FINGERPRINT_DOMAIN, canonicalPayeeJson(option)));
  }
  const sorted = dedupeAndSort(payeeFingerprints);
  const payeeSetFingerprint = await domainSeparatedHash(SET_FINGERPRINT_DOMAIN, sorted.join("\n"));
  return { version: FINGERPRINT_VERSION, payeeSetFingerprint, payeeFingerprints: sorted };
}
